ipa-core/src/query/runner/hybrid.rs

```diff
@@ -50,41 +49,51 @@ where
             key_registry,
             phantom_data: _,
         } = self;
+
         tracing::info!("New hybrid query: {config:?}");
-        let _ctx = ctx.narrow(&Hybrid);
+        let ctx = ctx.narrow(&Hybrid);
         let sz = usize::from(query_size);
 
-        let mut unique_encrypted_hybrid_reports = UniqueBytesValidator::new(sz);
-
         if config.plaintext_match_keys {
             return Err(Error::Unsupported(
                 "Hybrid queries do not currently support plaintext match keys".to_string(),
             ));
         }
 
-        let _input = LengthDelimitedStream::<EncryptedHybridReport, _>::new(input_stream)
-            .map_err(Into::<Error>::into)
-            .and_then(|enc_reports| {
-                future::ready(
-                    unique_encrypted_hybrid_reports
-                        .check_duplicates(&enc_reports)
-                        .map(|()| enc_reports)
-                        .map_err(Into::<Error>::into),
-                )
-            })
-            .map_ok(|enc_reports| {
-                iter(enc_reports.into_iter().map({
-                    |enc_report| {
-                        enc_report
-                            .decrypt::<R, BreakdownKey, Value, Timestamp>(key_registry.as_ref())
-                            .map_err(Into::<Error>::into)
-                    }
-                }))
-            })
-            .try_flatten()
-            .take(sz)
-            .try_collect::<Vec<_>>()
-            .await?;
+        let (_decrypted_reports, tags): (Vec<HybridReport<BA8, BA3>>, Vec<UniqueTag>) =
+            LengthDelimitedStream::<EncryptedHybridReport, _>::new(input_stream)
+                .map_err(Into::<Error>::into)
+                .map_ok(|enc_reports| {
+                    iter(enc_reports.into_iter().map({
+                        |enc_report| {
+                            let dec_report = enc_report
+                                .decrypt::<R, BA8, BA3, BA20>(key_registry.as_ref())
+                                .map_err(Into::<Error>::into);
+                            let unique_tag = UniqueTag::from_unique_bytes(&enc_report);
+                            dec_report.map(|dec_report1| (dec_report1, unique_tag))
+                        }
+                    }))
+                })
+                .try_flatten()
+                .take(sz)
+                .try_fold((Vec::new(), Vec::new()), |mut acc, result| async move {
+                    acc.0.push(result.0);
+                    acc.1.push(result.1);
+                    Ok(acc)
+                })
+                .await?;
+
+        let resharded_tags = reshard_iter(ctx.clone(), tags, |ctx, _, tag| {
+            tag.shard_picker(ctx.shard_count())
+        })
+        .await?;
```

Comment:
I want to look into this - I think it may be possible to unzip the stream into two and feed tag stream to resharding and keep collecting reports from a separate stream

Reply:
agreed. my thought was to land this that works, then give that another attempt (which will make that diff easier to review and also much easier to share what's not working if I get stuck.)